ostd/src/task/scheduler/mod.rs

```diff
@@ -0,0 +1,246 @@
+// SPDX-License-Identifier: MPL-2.0
+
+#![allow(dead_code)]
+
+mod fifo_scheduler;
+
+pub use fifo_scheduler::FifoScheduler;
+use spin::Once;
+
+use super::{processor::switch_to_task, task::Task, TaskStatus};
+use crate::{arch::timer::register_callback, prelude::*};
+
+static SCHEDULER: Once<&'static dyn Scheduler<Task>> = Once::new();
+
+/// Injects a scheduler implementation into framework.
+///
+/// This function can only be called once and must be called during the initialization of kernel.
+pub fn inject_scheduler(scheduler: &'static dyn Scheduler) {
+    SCHEDULER.call_once(|| scheduler);
+
+    register_callback(|| {
+        SCHEDULER.get().unwrap().local_mut_rq_with(&mut |local_rq| {
+            if local_rq.update_current(UpdateFlags::Tick) {
+                local_rq.set_should_preempt(true);
+            }
+        })
+    });
+}
+
+/// Adds task to scheduler.
+///
+/// This function may be called with newly spawned task or waked task.
+pub fn add_task(runnable: Arc<Task>, flags: EnqueueFlags) {
+    if !SCHEDULER.is_completed() {
+        let fifo_scheduler = Box::new(FifoScheduler::<Task>::new());
+        let static_scheduler = Box::leak(fifo_scheduler);
+        inject_scheduler(static_scheduler);
+    }
+    SCHEDULER.get().unwrap().enqueue(runnable, flags);
+}
+
+/// Yields execution.
+pub fn yield_now(flags: YieldFlags) {
+    reschedule(&mut |local_rq| {
```

Comment:
The code in this closure is hard to read. The current code mixed the logic dealing with Exit, Wait and Yield. I suggest dealing with one flag once to make the code more clear.

```rust
if flags == YieldFlags::Wait {
    // ......
}

if flags == YieldFlags::Yield {
    // ......
}

if flags == YieldFlags::Exit {
    //......
}
```

Further, it's better not to define local uninitialized variable. 
```rust
let current_task;
```

Reply:
OK, I'll split this function and possibly remove the `YieldFlags`.